optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/calendar/TwoDayView.java

```diff
@@ -0,0 +1,1013 @@
+import org.optaplanner.openshift.employeerostering.gwtui.client.common.Value;
+import org.optaplanner.openshift.employeerostering.gwtui.client.interfaces.HasTimeslot;
+import org.optaplanner.openshift.employeerostering.gwtui.client.popups.ErrorPopup;
+import org.optaplanner.openshift.employeerostering.shared.timeslot.TimeSlotUtils;
+
+public class TwoDayView<G extends HasTitle, I extends HasTimeslot<G>, D extends TimeRowDrawable<G>> implements
+        CalendarView<G,
+                I>, HasRows, HasData<
+                        Collection<D>> {
+
+    Calendar<G, I> calendar;
+
+    private static final String BACKGROUND_1 = "#efefef";
+    private static final String BACKGROUND_2 = "#e0e0e0";
+    private static final String LINE_COLOR = "#000000";
+    private static final String[] WEEKDAYS = LocaleInfo.getCurrentLocale().getDateTimeFormatInfo().weekdaysFull();
+    private static final int WEEK_START = LocaleInfo.getCurrentLocale().getDateTimeFormatInfo().firstDayOfTheWeek();
+
+    private static final int HEADER_HEIGHT = 64;
+    private static final double SPOT_NAME_WIDTH = 200;
+
+    private List<G> groups = new ArrayList<>();
+    private Collection<Handler> rangeHandlers = new ArrayList<>();
+    private Collection<com.google.gwt.view.client.RowCountChangeEvent.Handler> rowCountHandlers = new ArrayList<>();
+    private HashMap<G, Integer> groupPos = new HashMap<>();
+    private HashMap<G, Integer> groupIndexOf = new HashMap<>();
+    private HashMap<G, Integer> groupEndPos = new HashMap<>();
+    private HashMap<G, Integer> cursorIndex = new HashMap<>();
+    private HashMap<G, DynamicContainer> groupContainer = new HashMap<>();
+    private HashMap<G, DynamicContainer> groupAddPlane = new HashMap<>();
+    private Collection<I> shifts;
+    private NativeHorizontalScrollbar scrollBar;
+
+    TimeSlotTable<D, G> timeslotTable;
+
+    private Collection<D> shiftDrawables;
+    private List<Collection<D>> cachedVisibleItems;
+    private List<Collection<D>> allItems;
+    private Pagination pagination;
+    private SimplePager pager;
+    private ListDataProvider<Collection<D>> dataProvider = new ListDataProvider<>();
+    private SelectionModel<? super Collection<D>> selectionModel;
+    private int rangeStart, rangeEnd;
+    private int totalDisplayedSpotSlots;
+    LocalDateTime baseDate;
+    LocalDateTime currDay;
+    LocalDateTime hardStartDateBound;
+
+    @Override
+    public LocalDateTime getHardStartDateBound() {
+        return hardStartDateBound;
+    }
+
+    @Override
+    public void setHardStartDateBound(LocalDateTime hardStartDateBound) {
+        this.hardStartDateBound = hardStartDateBound;
+        if (getViewStartDate().isBefore(hardStartDateBound)) {
+            setDate(hardStartDateBound);
+        }
+        updateScrollBar();
+    }
+
+    @Override
+    public LocalDateTime getHardEndDateBound() {
+        return hardEndDateBound;
+    }
+
+    private void updateScrollBar() {
```

Comment:
The scrollbar is specific to `TwoDayView`, so it be odd for the presenter to hold methods that will only ever be used by a single view. I can probably move a few of these methods to an internal "ViewState" object which can be unit tested (since it would have no UI elements)